File: ga-graphai/packages/graphai/src/federation.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Sequence


@dataclass(frozen=True)
class FederatedDataNode:
    """Metadata describing an OSINT enclave in the federation."""

    node_id: str
    locality: str
    privacy_budget: float
    sensitivity_ceiling: int
    latency_penalty_ms: int
    supported_capabilities: tuple[str, ...]
    sovereign: bool = False

    def can_process(self, capability: str, sensitivity: int) -> bool:
        return (
            capability in self.supported_capabilities
            and sensitivity <= self.sensitivity_ceiling
        )


@dataclass(frozen=True)
class FederatedQueryRequest:
    query_id: str
    required_capabilities: tuple[str, ...]
    sensitivity: int
    preferred_localities: tuple[str, ...]
    privacy_budget: float
    estimated_edges: int


@dataclass(frozen=True)
class FederatedPlanStep:
    node_id: str
    capability: str
    estimated_latency_ms: int
    privacy_cost: float
    rationale: str
    secure_aggregation: bool


@dataclass(frozen=True)
class FederatedQueryPlan:
    query_id: str
    steps: tuple[FederatedPlanStep, ...]
    residual_budget: float
    warnings: tuple[str, ...]


class FederatedQueryPlanner:
    """Greedy planner that honours locality and privacy budgets."""

    def __init__(self, nodes: Sequence[FederatedDataNode]):
        self._nodes = list(nodes)

    def plan(self, request: FederatedQueryRequest) -> FederatedQueryPlan:
        if not request.required_capabilities:
            raise ValueError("at least one capability is required")
        steps: list[FederatedPlanStep] = []
        budget_remaining = request.privacy_budget
        warnings: list[str] = []

        ordered_nodes = sorted(
            self._nodes,
            key=lambda node: (
                0 if node.locality in request.preferred_localities else 1,
                node.latency_penalty_ms,
                -node.privacy_budget,
            ),
        )

        for capability in request.required_capabilities:
            candidate = self._select_node(ordered_nodes, capability, request.sensitivity)
            if candidate is None:
                warnings.append(
                    f"no node can satisfy capability {capability} at sensitivity {request.sensitivity}"
                )
                continue

            privacy_cost = max(0.05, request.estimated_edges / 100_000)
            if candidate.sovereign and candidate.locality not in request.preferred_localities:
                privacy_cost *= 1.5
            budget_remaining -= privacy_cost

            step = FederatedPlanStep(
                node_id=candidate.node_id,
                capability=capability,
                estimated_latency_ms=candidate.latency_penalty_ms + 25,
                privacy_cost=round(privacy_cost, 4),
                rationale=self._rationale(candidate, capability, request),
                secure_aggregation=candidate.sovereign,
            )
            steps.append(step)

        if budget_remaining < 0:
            warnings.append(
                "privacy budget exceeded; consider trimming query scope or rebalancing enclaves"
            )

        return FederatedQueryPlan(
            query_id=request.query_id,
            steps=tuple(steps),
            residual_budget=round(max(budget_remaining, 0.0), 4),
            warnings=tuple(warnings),
        )

    @staticmethod
    def _select_node(
        nodes: Iterable[FederatedDataNode], capability: str, sensitivity: int
    ) -> FederatedDataNode | None:
        for node in nodes:
            if node.can_process(capability, sensitivity):
                return node
        return None

    @staticmethod
    def _rationale(
        node: FederatedDataNode, capability: str, request: FederatedQueryRequest
    ) -> str:
        locality_hint = (
            "preferred locality match"
            if node.locality in request.preferred_localities
            else "cross-region placement"
        )
        return (
            f"route {capability} to {node.node_id} due to {locality_hint} and latency"
            f" allowance {node.latency_penalty_ms} ms"
        )

File: ga-graphai/packages/workcell-runtime/src/federation.ts
import type { PolicyEvaluationResult } from 'common-types';

export interface FederatedDataNode {
  id: string;
  locality: string;
  privacyBudget: number;
  sensitivityCeiling: number;
  latencyPenaltyMs: number;
  capabilities: string[];
  sovereign?: boolean;
}

export interface SubgraphQueryRequest {
  queryId: string;
  tenantId: string;
  requestedBy: string;
  roles: string[];
  requiredCapabilities: string[];
  preferredLocalities?: string[];
  sensitivity: number;
  estimatedEdges: number;
  privacyBudget: number;
  attributes?: Record<string, string | number | boolean>;
}

export interface SubgraphPlanStep {
  nodeId: string;
  capability: string;
  estimatedLatencyMs: number;
  privacyCost: number;
  locality: string;
  secureAggregation: boolean;
  rationale: string[];
  policy?: PolicyEvaluationResult;
}

export interface SubgraphPlan {
  queryId: string;
  tenantId: string;
  steps: SubgraphPlanStep[];
  residualBudget: number;
  warnings: string[];
}

export class FederatedPlanner {
  private readonly nodes: FederatedDataNode[] = [];

  registerNode(node: FederatedDataNode): void {
    this.nodes.push(node);
  }

  plan(request: SubgraphQueryRequest): SubgraphPlan {
    if (!request.requiredCapabilities.length) {
      throw new Error('at least one capability must be requested');
    }

    const preferred = new Set(request.preferredLocalities ?? []);
    const steps: SubgraphPlanStep[] = [];
    const warnings: string[] = [];
    let remainingBudget = request.privacyBudget;

    for (const capability of request.requiredCapabilities) {
      const candidate = this.selectNode(capability, request.sensitivity, preferred);
      if (!candidate) {
        warnings.push(
          `no node available for capability ${capability} within sensitivity ${request.sensitivity}`,
        );
        continue;
      }

      const baseCost = Math.max(0.05, request.estimatedEdges / 100_000);
      const sovereigntyMultiplier =
        candidate.sovereign && !preferred.has(candidate.locality) ? 1.5 : 1;
      const privacyCost = Number((baseCost * sovereigntyMultiplier).toFixed(4));
      remainingBudget -= privacyCost;

      const rationale = [
        preferred.has(candidate.locality)
          ? 'preferred locality match'
          : 'cross-region placement',
        `latency allowance ${candidate.latencyPenaltyMs}ms`,
      ];

      steps.push({
        nodeId: candidate.id,
        capability,
        estimatedLatencyMs: candidate.latencyPenaltyMs + 25,
        privacyCost,
        locality: candidate.locality,
        secureAggregation: Boolean(candidate.sovereign),
        rationale,
      });
    }

    if (remainingBudget < 0) {
      warnings.push('privacy budget exceeded for requested subgraph query');
    }

    return {
      queryId: request.queryId,
      tenantId: request.tenantId,
      steps,
      residualBudget: Number(Math.max(0, remainingBudget).toFixed(4)),
      warnings,
    };
  }

  private selectNode(
    capability: string,
    sensitivity: number,
    preferred: Set<string>,
  ): FederatedDataNode | undefined {
    const sorted = [...this.nodes].sort((a, b) => {
      const aPref = preferred.has(a.locality) ? 0 : 1;
      const bPref = preferred.has(b.locality) ? 0 : 1;
      if (aPref !== bPref) {
        return aPref - bPref;
      }
      if (a.latencyPenaltyMs !== b.latencyPenaltyMs) {
        return a.latencyPenaltyMs - b.latencyPenaltyMs;
      }
      return b.privacyBudget - a.privacyBudget;
    });

    return sorted.find(
      (node) =>
        node.capabilities.includes(capability) && sensitivity <= node.sensitivityCeiling,
    );
  }
}
